### ISSUE TITLE
Subdirectory .rubocop.yml with any rules besides inheritance causes some cops to fail

### ISSUE THREAD

**Issue by agrberg** [rubocop/rubocop#5576]
As part of a Rails 5 upgrade I've come across the new style for Rails controller testing. Thankfully this is covered and has autocorrect available in the `Rails/HttpPositionalArguments` Cop. Unfortunately this cop simply either does not run or does not work when I have a `spec/.rubocop.yml`. I've commented out the entire file and removed all the contents before saving and running rubocop again. Until I actually remove the file the cop fails.

I'm not sure how many other cops this problem extends to. The spec local .rubocop.yml is just used to change some linting, metric, and rubocop-rspec behaviors. I do not believe the contents matters as the presence of a blank file triggers the error as well.

--------

## Expected behavior

`rubocop --rails --only HttpPositionalArguments` should list ~150 failures in my test suite.

## Actual behavior

No offenses are detected

## Steps to reproduce the problem

If I simply do `touch spec/.rubocop.yml` before running rubocop I am not notified of the HttpPositionalArguments failures.

## RuboCop version

Include the output of `rubocop -V`. Here's an example:

```
$ rubocop -V
0.52.1 (using Parser 2.4.0.2, running on ruby 2.5.0 x86_64-darwin16)
```

**Comment by agrberg** [comment]
## Workaround

Since my primary goal was to fix these errors as a part of my upgrade the following workaround worked for me:

```bash
rm spec/.rubocop.yml
rubocop --rails --only HttpPositionalArguments --auto-correct
git checkout HEAD -- spec/.rubocop.yml
```

I hope this helps anyone who finds this issue or a similar one.

**Comment by pocke** [comment]
A `.rubocop.yml` in a sub directory conceals the parent .rubocop.yml.

Minimum reproduce
===


```bash
$ vim .rubocop.yml
Rails:
  Enabled: true

$ vim test.rb
puts 'it should be a.present?' unless a.blank?

$ mkdir a
$ cp test.rb a/test2.rb

$ rubocop 
Inspecting 2 files
CC

Offenses:

test.rb:1:32: C: Rails/Present: Use if a.present? instead of unless a.blank?.
puts 'it should be a.present?' unless a.blank?
                               ^^^^^^^^^^^^^^^
a/test2.rb:1:32: C: Rails/Present: Use if a.present? instead of unless a.blank?.
puts 'it should be a.present?' unless a.blank?
                               ^^^^^^^^^^^^^^^

2 files inspected, 2 offenses detected

$ touch a/.rubocop.yml

$ rubocop 
Inspecting 2 files
C.

Offenses:

test.rb:1:32: C: Rails/Present: Use if a.present? instead of unless a.blank?.
puts 'it should be a.present?' unless a.blank?
                               ^^^^^^^^^^^^^^^

2 files inspected, 1 offense detected

```


I'm not sure it is a bug or not.


BTW, we can use `inherit_from` to avoid this problem.


```yaml
# a/.rubocop.yml
inherit_from: ../.rubocop.yml
```

```bash
$ rubocop 
Inspecting 2 files
CC

Offenses:

test.rb:1:32: C: Rails/Present: Use if a.present? instead of unless a.blank?.
puts 'it should be a.present?' unless a.blank?
                               ^^^^^^^^^^^^^^^
a/test2.rb:1:32: C: Rails/Present: Use if a.present? instead of unless a.blank?.
puts 'it should be a.present?' unless a.blank?
                               ^^^^^^^^^^^^^^^

2 files inspected, 2 offenses detected
```

**Comment by agrberg** [comment]
Thank you for clarifying that a `.rubocop.yml` in a subdirectory will conceal the parent one. That was not clear to me and I can confirm that the failures appear when my `spec/.rubocop.yml` consists only of `inherit_from: ../.rubocop.yml`. The reason I reported that the mere presence of the file caused problems is because when it wouldn't report the errors, I comment out the whole file before removing all of the contents.

I suppose the _actual_ problem I'm experiencing is that _any_ additional content beyond the inheritance causes the errors to not be reported by Rubocop. For example, I have `TargetRailsVersion: 5.0` in `.rubocop.yml`. Then `spec/.rubocop.yml` looks like the following

```yml
inherit_from: ../.rubocop.yml

Metrics/LineLength:
  Max: 170
```

With just that minor addition, Rubocop no longer reports the `Rails/HttpPositionalArguments` errors.

**Comment by bbatsov** [comment]
I'm closing this ticket due to no recent activity. Feel free to re-open it if you ever come back to it.

**Comment by zenspider** [comment]
> BTW, we can use inherit_from to avoid this problem.

No, you can't.

@agrberg's findings are correct. ANYTHING besides `inherit_from` blows up and disables the config. I just got bit by this and it was an utter bitch to diagnose. I ran with `--debug` and eventually thought "what if I delete that sub-config?" and eventually found myself here.

This should be reopened... `inherit_from` and friends do NOT behave the way one would expect. My guess is the sub-config is inheriting from... but then the config after that is overwriting, not merging.

**Comment by zenspider** [comment]
I should add... I tried to be clever and chain some inherit_from files to make this work. 2 inherit_froms... etc... none worked. Making subdir configs simply buggy at this point.

**Comment by bbatsov** [comment]
@dvandersluis Can you take a look into this one, since you've been on a roll with config-related isssue lately? 😆 I planned to look into this one myself, but it's a bit crazy on my end this week.

**Comment by dvandersluis** [comment]
What is the expected behaviour here? Does [inherit_mode](https://docs.rubocop.org/rubocop/1.6/configuration.html#merging-arrays-using-inherit_mode) help?

**Comment by bbatsov** [comment]
> A .rubocop.yml in a sub directory conceals the parent .rubocop.yml.

Originally a `.rubocop.yml` in a subdirectory was supposed to override settings in a config file higher up the directory tree for any files under the directory it was placed in (e.g. `lib`, `app`, whatever). This functionality was created so you can have different configuration for different parts of the code - I think the use-case that made us develop this back in the day was that you don't care as much about some checks for specs (e.g. complexity) as you'd care for them in application/library code, so you'd place a more relaxed `.rubocop.yml` in your `spec` folder. I don't think anyone should ever need `inherit_from` for this particular functionality, so I'm guessing we broke something along the way. That's a little-known feature of RuboCop (but a pretty cool one) and I assume that not many people are using it.

**Comment by dvandersluis** [comment]
So the [docs](https://docs.rubocop.org/rubocop/1.6/configuration.html#config-file-locations) lay out how the config file is found, cascading up from the current directory until a file is found (" RuboCop will use the config as specified inside the **first** of the following files"). Are we looking to change this? Because in order for a config file to override a higher file, all would need to be loaded.

I think this might be a bigger change than anticipated, and cause some tricky behaviour, for instance: a config file above the project root will now apply to a project but will not be present for other developers. Theoretically then we can end up with nondeterministic behaviour where rubocop reports different offenses (or even conflicting offenses!) for different developers of the same project.

There's also the weird edge case that (I think?) `rubocop spec/` and `cd spec && rubocop` would have different rules applied to files in `spec` if `spec/.rubocop.yml` exists. Maybe having rubocop look _down_ instead of up would make sense here; however we currently have no mechanism to apply a nested config only to that directory, and also this would probably end up spending a lot of time scanning directories for configs.

**Comment by dvandersluis** [comment]
Also I am so far unable to replicate the bug:

```yaml
# .rubocop.yml

require:
  - rubocop-rails

Rails:
  Enabled: true
```

```yaml
# spec/.rubocop.yml

inherit_from: '../.rubocop.yml'

Rails/Pick:
  Enabled: false
```

```shell
$ rubocop spec/test_spec.rb --only Rails/Present
Inspecting 1 file
C

Offenses:

spec/test_spec.rb:1:9: C: Rails/Present: Use if a.present? instead of unless a.blank?.
  a = b unless a.blank?
        ^^^^^^^^^^^^^^^

1 file inspected, 1 offense detected, 1 offense auto-correctable
```

```shell
$ cd spec
$ rubocop test_spec.rb --only Rails/Present

Offenses:

test_spec.rb:1:9: C: Rails/Present: Use if a.present? instead of unless a.blank?.
  a = b unless a.blank?
        ^^^^^^^^^^^^^^^

1 file inspected, 1 offense detected, 1 offense auto-correctable
```

@zenspider can you make a minimal repo that demonstrates the problem?

**Comment by bbatsov** [comment]
> So the docs lay out how the config file is found, cascading up from the current directory until a file is found (" RuboCop will use the config as specified inside the first of the following files", emphasis mine). Are we looking to change this? Because in order for a config file to override a higher file, all would need to be loaded.

Hmm, seems I didn't remember how this was supposed to work - somehow I thought it was supposed to automatically merge the with the config file above it, but I guess that's not the case. At least not anymore.

**Comment by zenspider** [comment]
I don't think it needs to be changed. `inherit_from` should be sufficient for 90% of the projects out there that want to do something like this. 

@dvandersluis key to this in in the title: _causes *some* cops to fail_. Please try a repro with `Rails/HttpPositionalArguments` instead.

**Comment by vanboom** [comment]
Starting my first attempt with rubocop, I placed a .rubocop.yml file in the ./spec directory.  After running rubocop, I would get a lot of offenses detected but Rails/HttpPositionalArguments would return no offenses.  

Moving the .rubocop.yml file to the root directory resolved this behavior.  Thanks for the details here - helped me isolate this behavior.

**Comment by jonas054** [comment]
It might not be obvious, so I should explain why the PR #9706 solves the problem with `Rails/HttpPositionalArguments`.

It's because `spec/.rubocop.yml` inherits from `.rubocop.yml` one level up and that file inherits from `config/default.yml` in the `rubocop-rails` gem. The [configuration there](https://github.com/rubocop/rubocop-rails/blob/v2.9.1/config/default.yml#L340-L346) has `spec/**/*` as a path in the `Include` parameter. This path is inherited as-is to the `.rubocop.yml` on the base level and then down to `spec/.rubocop.yml`. And that's a problem. Include paths are relative to the directory where the configuration file is, making the path `spec/spec/**/*` effectively, and that doesn't match any files.
